cpp/src/arrow/compute/kernels/take-test.cc

```diff
@@ -62,82 +67,111 @@ class TestTakeKernel : public ComputeFixture, public TestBase {
-                                       "[0, 9, 0]", options, &arr));
+  ASSERT_RAISES(IndexError,
+                this->Take(boolean(), "[true, false, true]", int8(), "[0, 9, 0]", &arr));
 }
 
 template <typename ArrowType>
 class TestTakeKernelWithNumeric : public TestTakeKernel<ArrowType> {
  protected:
   void AssertTake(const std::string& values, const std::string& indices,
-                  TakeOptions options, const std::string& expected) {
-    TestTakeKernel<ArrowType>::AssertTake(type_singleton(), values, indices, options,
-                                          expected);
+                  const std::string& expected) {
+    TestTakeKernel<ArrowType>::AssertTake(type_singleton(), values, indices, expected);
   }
   std::shared_ptr<DataType> type_singleton() {
     return TypeTraits<ArrowType>::type_singleton();
   }
+  void ValidateTake(const std::shared_ptr<Array>& values,
+                    const std::shared_ptr<Array>& indices_boxed) {
+    std::shared_ptr<Array> taken;
+    TakeOptions options;
+    ASSERT_OK(
+        arrow::compute::Take(&this->ctx_, *values, *indices_boxed, options, &taken));
+    ASSERT_EQ(indices_boxed->length(), taken->length());
+
+    ASSERT_EQ(indices_boxed->type_id(), Type::INT32);
+    auto indices = checked_pointer_cast<Int32Array>(indices_boxed);
+    for (int64_t i = 0; i < indices->length(); ++i) {
+      if (indices->IsNull(i)) {
+        ASSERT_TRUE(taken->IsNull(i));
+        continue;
+      }
+      int32_t taken_index = indices->Value(i);
+      ASSERT_TRUE(values->RangeEquals(taken_index, taken_index + 1, i, taken));
+    }
+  }
 };
 
 TYPED_TEST_CASE(TestTakeKernelWithNumeric, NumericArrowTypes);
 TYPED_TEST(TestTakeKernelWithNumeric, TakeNumeric) {
-  TakeOptions options;
-  this->AssertTake("[7, 8, 9]", "[0, 1, 0]", options, "[7, 8, 7]");
-  this->AssertTake("[null, 8, 9]", "[0, 1, 0]", options, "[null, 8, null]");
-  this->AssertTake("[7, 8, 9]", "[null, 1, 0]", options, "[null, 8, 7]");
-  this->AssertTake("[null, 8, 9]", "[]", options, "[]");
+  this->AssertTake("[7, 8, 9]", "[0, 1, 0]", "[7, 8, 7]");
+  this->AssertTake("[null, 8, 9]", "[0, 1, 0]", "[null, 8, null]");
+  this->AssertTake("[7, 8, 9]", "[null, 1, 0]", "[null, 8, 7]");
+  this->AssertTake("[null, 8, 9]", "[]", "[]");
 
   std::shared_ptr<Array> arr;
   ASSERT_RAISES(IndexError, this->Take(this->type_singleton(), "[7, 8, 9]", int8(),
-                                       "[0, 9, 0]", options, &arr));
+                                       "[0, 9, 0]", &arr));
+}
+
+TYPED_TEST(TestTakeKernelWithNumeric, TakeRandomNumeric) {
+  auto rand = random::RandomArrayGenerator(kSeed);
+  for (size_t i = 3; i < 8; i++) {
+    const int64_t length = static_cast<int64_t>(1ULL << i);
+    for (size_t j = 0; j < 13; j++) {
+      const int64_t indices_length = static_cast<int64_t>(1ULL << j);
+      for (auto null_probability : {0.0, 0.01, 0.1, 0.25, 0.5, 1.0}) {
```

Comment:
Don't think it's useful to test so many probabilities...

Reply:
I can thin those out, then